R/kenwardroger.R

```diff
@@ -0,0 +1,198 @@
+#' Obtain Kenward-Roger Adjustment Components
+#'
+#' @param tmb_data (`mmrm_tmb_data`)\cr produced by [h_mmrm_tmb_data()].
+#' @param theta (`numeric`)\cr theta estimate.
+#'
+#' @details the function returns a named list, \eqn{P}, \eqn{Q} and \eqn{R}, which corresponds to the
+#' paper in 1997. The matrices are stacked in columns so that \eqn{P}, \eqn{Q} and \eqn{R} has the same
+#' column number(number of beta parameters). The number of rows, is dependent on
+#' the total number of theta and number of groups, if the fit is a grouped mmrm.
+#' For \eqn{P} matrix, it is stacked sequentially. For \eqn{Q} and \eqn{R} matrix, it is stacked so
+#' that the \eqn{Q_{ij}} and \eqn{R_{ij}} is stacked from \eqn{j} then to \eqn{i}, i.e. \eqn{R_{i1}}, \eqn{R_{i2}}, etc.
+#' \eqn{Q} and \eqn{R} only contains intra-group results and inter-group results should be all zero matrices
+#' so they are not stacked in the result.
+#'
+#' @return Named list with elements:
+#' - `P`: `matrix` of \eqn{P} component.
+#' - `Q`: `matrix` of \eqn{Q} component.
+#' - `R`: `matrix` of \eqn{R} component.
+#'
+#' @keywords internal
+h_get_kr_comp <- function(tmb_data, theta) {
+  .Call(`_mmrm_get_pqr`, PACKAGE = "mmrm", tmb_data, theta)
+}
+
+
+#' Calculation of Kenward-Roger Degrees of Freedom for Multi-Dimensional Contrast
+#'
+#' @param object (`mmrm`)\cr object.
+#' @param contrast (`matrix`)\cr contrast matrix.
+#' @param linear (`logical`)\cr whether to use linear Kenward-Roger approximation.
+#'
+#' @return List with `num_df`, `denom_df`, `f_stat` and `p_val` (2-sided p-value).
+#'
+#' @keywords internal
+h_df_md_kr <- function(object, contrast, linear = FALSE) {
+  if (component(object, "reml") != 1) {
```

Comment:
please also here minimum assertions

Reply:
in `df_1d` and `df_md` there are already assertions. shall we still duplicate the assertions here (and also h_df_md_sat) ?

Reply:
Basically in every function we need to assert the inputs that we use directly in this function. If arguments are just passed on to other function calls, no need to assert them in this function. No worries about speed when duplicating assertions, as checkmate is very fast, so that is not a concern.